symfony/plugins/orangehrmRESTPlugin/lib/Api/User/EmployeePunchInAPI.php

```diff
@@ -0,0 +1,166 @@
+     * @throws RecordNotFoundException
+     */
+    public function savePunchIn()
+    {
+        $timeZone = $this->getRequestParams()->getPostParam(parent::PARAMETER_TIME_ZONE);
+        $punchInNote = $this->getRequestParams()->getPostParam(parent::PARAMETER_NOTE);
+        $dateTime = $this->getRequestParams()->getPostParam(parent::PARAMETER_DATE_TIME);
+        if(empty($dateTime)) {
+            throw new InvalidParamException('Datetime Cannot Be Empty');
+        }
+        $empNumber = \sfContext::getInstance()->getUser()->getAttribute("auth.empNumber");
+        if($this->checkValidEmployee($empNumber)){
+            $actionableStatesList = array(\PluginAttendanceRecord::STATE_PUNCHED_IN);
+            $attendanceRecord = $this->getAttendanceService()->getLastPunchRecord($empNumber, $actionableStatesList);
+            if (is_null($attendanceRecord)) {
+                $attendanceRecord = new \AttendanceRecord();
+                $attendanceRecord->setEmployeeId($empNumber);
+
+                $nextState = \PluginAttendanceRecord::STATE_PUNCHED_IN;
+                if($timeZone){
+                    $zoneList = timezone_identifiers_list();
+                    if(in_array($timeZone,$zoneList)) {
+                        $timeZone_dtz = new \DateTimeZone($timeZone);
+                        $origin_dt = new \DateTime($dateTime, $timeZone_dtz);
+                        $punchIndateTime = $origin_dt->format('Y-m-d H:i');
+                        $timeZoneOffset = $this->getTimezoneOffset('UTC', $timeZone);
+                        //check overlapping
+                        $punchInUtcTime= Date('Y-m-d H:i:s', strtotime($punchIndateTime) + ((-1) * $timeZoneOffset));
+                        $isValid = $this->getAttendanceService()->checkForPunchInOverLappingRecords($punchInUtcTime,$empNumber);
+                        if(!$isValid){
+                            throw new InvalidParamException('Overlapping Records Found');
+                        }
+
+                    }else{
+                        throw new InvalidParamException('Invalid Time Zone');
+                    }
+                }else {
+                    $punchIndateTime = date($dateTime,'Y-m-d H:i');
+                    $timeZoneOffset = $this->getTimezoneOffset('UTC');
+                }
+                try {
+                    $attendanceRecord = $this->setPunchInRecord(
+                        $attendanceRecord,
+                        $nextState,
+                        $punchInUtcTime,
+                        $punchIndateTime,
+                        $timeZoneOffset / 3600,
+                        $punchInNote
+                    );
+                    $origin_dtz = new \DateTimeZone($timeZone);
+                    $origin_dt = new \DateTime("now", $origin_dtz);
+                    $originTimeZoneOffset=$origin_dtz->getOffset($origin_dt)/3600;
+                    return new Response(array('success' => 'Successfully Punched In',
```

Comment:
Better to create a Model for the response

Reply:
not done yet